### PR TITLE
feature(datepicker): let users define his own date model implementation

### PR DESCRIPTION
**Pull request by edriang**
Before submitting a pull request, please make sure you have at least performed the following:

 - [] read and followed the [CONTRIBUTING.md](https://github.com/ng-bootstrap/ng-bootstrap/blob/master/CONTRIBUTING.md) guide.
 - [] built and tested the changes locally.
 - [] added/updated any applicable tests.
 - [] added/updated any applicable API documentation.
 - [] added/updated any applicable demos.

### CODE REVIEW

File: demo/src/app/components/datepicker/demos/adapter/datepicker-adapter.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import {Component} from '@angular/core';
+import {NgbDateStruct, NgbDateAdapter, NgbDate} from '@ng-bootstrap/ng-bootstrap';
+
+const now = new Date();
+
+/**
+ * Example of a Native Date adapter
+ */
+class NgbDateNativeAdapter extends NgbDateAdapter<Date> {
```
Comment (maxokorokov) [inline demo/src/app/components/datepicker/demos/adapter/datepicker-adapter.ts]: forgot `export class` and `@Injectable()`

File: demo/src/app/components/datepicker/demos/adapter/datepicker-adapter.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import {Component} from '@angular/core';
+import {NgbDateStruct, NgbDateAdapter, NgbDate} from '@ng-bootstrap/ng-bootstrap';
```
Comment (maxokorokov) [inline demo/src/app/components/datepicker/demos/adapter/datepicker-adapter.ts]: no need in importing `NgbDateStruct` here

File: src/datepicker/ngb-date.ts [hunk 3]
```diff
@@ -1,3 +1,6 @@
+import {Injectable, Inject, forwardRef} from '@angular/core';
+import {NgbDateAdapter} from './ngb-date-adapter';
```
Comment (maxokorokov) [inline src/datepicker/ngb-date.ts]: no need in any new imports in `ngb-date.ts`

File: src/datepicker/ngb-date-parser-formatter.ts [hunk 4]
```diff
@@ -1,5 +1,6 @@
 import {padNumber, toInteger, isNumber} from '../util/util';
 import {NgbDateStruct} from './ngb-date-struct';
+import {NgbDate} from './ngb-date';
```
Comment (maxokorokov) [inline src/datepicker/ngb-date-parser-formatter.ts]: no need in any new imports in `ngb-date-parser-formatter.ts`

File: src/datepicker/ngb-date-adapter.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
+import {NgbDateStruct} from './ngb-date-struct';
+import {NgbDate} from './ngb-date';
+
+/**
+ * Abstract type serving as a DI token for the service converting from your application Date model to internal NgbDate
+ * model.
+ * A default implementation converting from and to NgbDateStruct is provided for retro-compatibility,
+ * but you can provide another implementation to use an alternative format, ie for using with native Date Object.
+ */
+export abstract class NgbDateAdapter<T> {
```
Comment (maxokorokov) [inline src/datepicker/ngb-date-adapter.ts]: @Injectable()

File: src/datepicker/ngb-date-adapter.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  /**
+   * Converts internal date value NgbDate to user-model date
+   * The returned type is suposed to be of the same type as fromModel() input-value param
+   * @param  {NgbDate} date internal NgbDate date representation
+   * @return {any}
+   */
+  abstract toModel(date: NgbDate): T;
+}
+
+export class NgbDateStructAdapter extends NgbDateAdapter<NgbDateStruct> {
```
Comment (maxokorokov) [inline src/datepicker/ngb-date-adapter.ts]: @Injectable()

**Comment by SauceCode84** [comment]
Has this been resolved?

**Comment by edriang** [comment]
@SauceCode84 sorry, I will try to take a look on maxokorokov comments to fix the problem. Maybe this afternoon.

**Comment by edriang** [comment]
@maxokorokov Thank you very much, all your corrections worked perfectly. Sorry for the delay.

**Comment by eddieklc** [comment]
Hey guys, do you maybe have an estimate when is this upgrade going to be merged with the master branch and released?

**Comment by msosa** [comment]
this whole PR seems very useful, @edriang can you make the requested changes?

**Comment by adrian-gallardo-wizeline** [comment]
@msosa , I'm not longer working for the company which needed this change. Can anybody else continue with this development? you can fork the branch if you want. Thanks

**Pull request closed without merging**